backend/src/api.ts

```diff
@@ -0,0 +1,684 @@
+        this API call will take some time to finish. The 'type' must be either
+        "reference" or "candidate".`
+    },
+    {
+      path: "/api/test-records",
+      description: "Returns all of the tests results as a JSON object"
+    },
+    {
+      path: "/api/diff-threshold",
+      description: "Returns the current difference threshold used for image comparisons"
+    },
+    {
+      path: "/api/update-diff-threshold",
+      description: `(Requires admin) This will recalculate all of the difference images
+        with the new threshold that is passed to this function. The JSON object in the
+        body must contain the 'threshold' value as a number between 0 and 1 that is the
+        new error threshold`
+    },
+    {
+      path: "/api/submit-test",
+      description: `(Requires runner) This will submit a new test to the image testing
+        server. The body of message must contain a 'runnerID', 'hardware', 'group',
+        'name', 'timestamp', 'timing', and 'commitHash'. The 'runnerID' must be one of the
+        allowed runners setup for this server. Furthermore, there needs to be the
+        candidate file as a multipart encoded file`
+    },
+    {
+      path: "/api/run-test",
+      description: `This will run a comparison against the current reference image on the
+        server without storing the results. The body of message must contain a 'hardware',
+        'group', and 'name'. Furthermore, there needs to be the candidate file as a
+        multipart encoded file. The API returns the difference image.`
+    },
+    {
+      path: "/api/update-reference",
+      description: `(Requires admin) Marks the current reference image for a specific test
+        as invalid and instead use the latest test as a reference instead. In addition to
+        the admin token, the JSON object contained in the body must provide the
+        'hardware', 'group', and 'name' for test whose reference should be invalidated.`
+    }
+  ]);
+}
+
+
+
+/**
+ * Returns a single requested result. The URL parameters used for this function are:
+ *  - `type`: One of "reference", "candidate", "difference", "reference-thumbnail",
+ *            "candidate-thumbnail", "difference-thumbnail", or "log"
+ *  - `group`: The name of the test's group for which to return the image
+ *  - `name`: The name of the test for which to return the image
+ *  - `hardware`: The test's hardware for which to return the image
+ *  - `timestamp`: The timestamp of the test for which to return the image. This is an
+ *                 optional parameter. If it is left out, the latest result for the
+ *                 specified type will be returned
+ */
+async function handleResult(req: express.Request, res: express.Response) {
+  const types = [
+    "reference", "candidate", "difference", "reference-thumbnail", "candidate-thumbnail",
+    "difference-thumbnail", "log"
+  ] as const;
+
+  const p: any = req.params;
```

Comment:
I tried but its gets really wordy to have any proper type information in here.  Since the `type` would have to be of `typeof types` and then the `req.params` would have to be properly parsed which would result in worse error messages.

But if there is a better way for this, I'd be happy to see and have it